### PR TITLE
ahk_x11: init at 1.0.4

### LINKED ISSUE #279323: Package request: AHK_X11
**Project description**

AutoHotkey for Linux. Distributed as source, deb, and appimage.

**Metadata**

* homepage URL: https://github.com/phil294/AHK_X11
* source URL: https://github.com/phil294/AHK_X11
* license: GPL-2.0
* platforms: linux (maybe others, if running X11)

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by UlyssesZh**
## Description of changes

Adds package [AHK_X11](https://github.com/phil294/AHK_X11), an [AutoHotkey](https://www.autohotkey.com) implementation for Linux.

Closes #279323. ~~Closes #297972 (this PR contains the commit in that PR)~~ (already merged). CC @AndersonTorres.

~~It reports command not found for `file` command if your PATH does not contain that command. This error does not affect normal usage. I could have fixed by adding `file` as a dependency, but this is actually a non-fatal upstream bug (phil294/AHK_X11#75), so I left it like this~~ (Fixed in upstream).

AHK_X11 from this package does not have functionality of compiling scripts into portable executable. The implementation of that compilation feature is simply to concatenate the binary and the script to form an executable file, but executable files in NixOS are not executable in other systems. I was thinking packaging a `ahk_x11-bin` package for that demand, but I didn't do that because the executable cannot run in the environment where it was built (even with `steam-run` due to the requirement of `libnotify`).

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+, openbox
+, xvfb-run
+, at-spi2-atk
+
+, buildDevTarget ? false # the dev version prints debug info
+} :
+
+let
+  pname = "ahk_x11";
+  version = "1.0.4-b76359f"; # 1.0.4 hasn't been released
```
Comment (sikmir) [inline pkgs/by-name/ah/ahk_x11/package.nix:23]: See https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming
```suggestion
  version = "1.0.3-unstable-2024-04-22";
```

Comment (sikmir) [reply pkgs/by-name/ah/ahk_x11/package.nix:23]: Any reason to use latest instead of 1.0.3?

Comment (UlyssesZh) [reply pkgs/by-name/ah/ahk_x11/package.nix:23]: 1.0.3 uses an old version of gi-crystal which generates bad bindings with new libraries. This leads to build failure. I noted this in the comments above `rev`.

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+  inherit (xorg) libXinerama libXtst libXext libXi;
+
+in crystal.buildCrystalPackage {
+
+  inherit pname version;
+  src = fetchFromGitHub {
+    owner = "phil294";
+    repo = "AHK_X11";
+    # 1.0.3 doesn't build
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:38]:
```suggestion
```

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+, openbox
+, xvfb-run
+, at-spi2-atk
+
+, buildDevTarget ? false # the dev version prints debug info
+} :
+
+let
+  pname = "ahk_x11";
+  version = "1.0.3-unstable-2024-03-22"; # 1.0.4 hasn't been released
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:23]:
```suggestion
  version = "1.0.3-unstable-2024-03-22"; # 1.0.3 doesn't build
```

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+  meta = with lib; {
+    description = "AutoHotkey for X11";
+    homepage = "https://phil294.github.io/AHK_X11";
+    license = licenses.gpl2;
+    maintainers = with maintainers; [ ulysseszhan ];
+    platforms = platforms.linux;
+    mainProgram = "ahk_x11";
+  };
+
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:98]:
```suggestion
```

**Review by AndersonTorres — APPROVED** [review]
Nice bug catch!

**Comment by UlyssesZh** [comment]
The new push just now utilized the `copyShardDeps` flag introduced in #289266 (merged days ago).

#305484 may change the situation a bit, too, because this package is using gi-crystal. In https://github.com/NixOS/nixpkgs/pull/305484#issuecomment-2067747412, I was hoping we can have a better way of packaging Crystal packages that use gi-crystal.

CC @sund3RRR.

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  crystalLib = linkFarm "ahk_x11-lib" (lib.mapAttrsToList (name: value: {
+    inherit name;
+    path = fetchgit value;
+  }) (import ./shards.nix));
+
+  inherit (xorg) libXinerama libXtst libXext libXi;
+
+in crystal.buildCrystalPackage {
+
+  inherit pname version;
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:34]:
```suggestion
  inherit pname version;

```

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  src = fetchFromGitHub {
+    owner = "phil294";
+    repo = "AHK_X11";
+    rev = "047a89722f9680b2b753dee59e74a2d05e746849"; # version;
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:38]: version does not work here?

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    hash = "sha256-wVKHut1Z3kNVOmkBvQ3X1rW31DH1nL3tN3fED8OiOdM=";
+    fetchSubmodules = true;
+  };
+
+  postPatch = ''
+    sed -i 's/shards install --frozen//' Makefile
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:44]: use substitute* functions

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+  buildTargets = if buildDevTarget then "bin/ahk_x11.dev" else "bin/ahk_x11";
+  checkTarget = "test-dev";
+
+  # The tests fail with AtSpi failure. This means it lacks assistive technologies:
+  # https://github.com/phil294/AHK_X11?tab=readme-ov-file#accessibility
+  # I don't know how to fix it for xvfb and openbox.
+  doCheck = false;
+
+  meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:78]:
```suggestion
  meta = {
```

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    description = "AutoHotkey for X11";
+    homepage = "https://phil294.github.io/AHK_X11";
+    license = licenses.gpl2;
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:81]:
```suggestion
    license = lib.licenses.gpl2Plus;
```

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+    maintainers = with maintainers; [ ulysseszhan ];
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:82]:
```suggestion
    maintainers = with lib.maintainers; [ ulysseszhan ];
```

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    platforms = platforms.linux;
```
Comment (AndersonTorres) [inline pkgs/by-name/ah/ahk_x11/package.nix:83]:
```suggestion
    platforms = lib.platforms.linux;
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/ah/ahk_x11/package.nix (hunk 6 above)
Comment (UlyssesZh) [reply pkgs/by-name/ah/ahk_x11/package.nix:38]: See the comment following the definition of `version`. This is an untagged version. The latest tagged version does not build.

Comment (AndersonTorres) [reply pkgs/by-name/ah/ahk_x11/package.nix:33]: Delete the confusing comment

```suggestion
    rev = "047a89722f9680b2b753dee59e74a2d05e746849";
```

File: pkgs/by-name/ah/ahk_x11/package.nix [hunk 12]
```diff
@@ -0,0 +1,84 @@
+{ lib
+, stdenv
```
Comment (FliegendeWurst) [inline pkgs/by-name/ah/ahk_x11/package.nix:2]: Please format with the new formatter.

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**